### PR TITLE
Add some pattern-matching tests suggested by test plan.

### PR DESCRIPTION
**Pull request by gafter**
Relates to #19038, #16195

Note that this is a test-only change. Comments referring to added tests have been added to #19038.

@AlekseyTs Can you review this test-only change please?

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs [hunk 1]
```diff
@@ -5895,5 +5934,148 @@ public static void Main(string[] args)
...
+                Diagnostic(ErrorCode.ERR_PatternWrongGenericTypeInVersion, "Packet<T>").WithArguments("T", "Packet<T>", "7", "7.1").WithLocation(12, 52)
+                );
+            var compilation = CreateStandardCompilation(program, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular7_1);
+            compilation.VerifyDiagnostics(
+                );
+            var comp = CompileAndVerify(compilation, expectedOutput: @"FalseTrueFalseFalse");
+        }
+
+        [Fact, WorkItem(19038, "https://github.com/dotnet/roslyn/issues/19038")]
+        public void MatchRestrictedTypes()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: >MatchRestrictedTypes [](start = 20, length = 20)

I think we should verify runtime behavior of success cases, consider separating them into a dedicated unit-test.

**Comment by AlekseyTs** [comment]
Done with review pass.

**Comment by gafter** [comment]
retest please

**Comment by gafter** [comment]
retest prtest please

**Comment by gafter** [comment]
retest windows_debug_unit64_prtest please
retest windows_debug_vs-integration_prtest please
retest windows_release_vs-integration_prtest please

**Comment by gafter** [comment]
retest windows_debug_unit64_prtest please

**Comment by gafter** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by gafter** [comment]
retest windows_release_vs-integration_prtest please

**Comment by gafter** [comment]
@AlekseyTs I've added the requested test.

**Comment by gafter** [comment]
retest windows_release_unit32_prtest please

**Comment by gafter** [comment]
@AlekseyTs I've added the requested test. Do you have any other comments?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a review of this test-only change?

**Pull request merged**